✨ Add exemplars to prometheus metrics

bisgaard-itis opened the pull request:
<!-- Title Annotations:

  WIP: work in progress
  🐛    Fix a bug.
  ✨    Introduce new features.
  🎨    Enhance existing feature.
  ♻️    Refactor code.
  🚑️    Critical hotfix.
  ⚗️    Perform experiments.
  ⬆️    Upgrade dependencies.
  📝    Add or update documentation.
  🔨    Add or update development scripts.
  ✅    Add, update or pass tests.
  🔒️    Fix security issues.
  ⚠️    Changes in ops configuration etc. are required before deploying.
        [ Please add a link to the associated ops-issue or PR, such as in https://github.com/ITISFoundation/osparc-ops-environments or https://git.speag.com/oSparc/osparc-infra ]
  🗃️    Database table changed (relevant for devops).
  👽️    Public API changes (meaning: dev features are moved to being exposed in production)
  🚨    Do manual testing when deployed

or from https://gitmoji.dev/
-->

## What do these changes do?
- Add Prometheus [exemplars](https://grafana.com/docs/grafana/latest/fundamentals/exemplars/) to our Prometheus metrics. This will allow us (in a couple of mouse clicks) to navigate from the Grafana dashboard associated with a given simcore service to the trace of a **relevant** request. I.e. if we see a that response latency is very slow, we can navigate to the trace of one of the requests for which the server is so slow to respond. Since our logs are also instrumented we can use the trace to navigate to all the logs emitted during handling that request. One can then to the same for another similar request (same endpoint) at a point in time when response latency was not so slow and compare the traces/logs. This workflow is inspired by https://github.com/blueswen/fastapi-observability
- Since https://github.com/trallnag/prometheus-fastapi-instrumentator doesn't support adding exemplars and we have already hit the [limitations](https://github.com/trallnag/prometheus-fastapi-instrumentator/issues/317) of that library I decided to consolidate the aiohttp and fastapi middleware we use to collect prometheus metrics. Hence, I will also update our Grafana dashboards accordingly.

<!-- Badge to openapi specs
[![ReDoc](https://img.shields.io/badge/OpenAPI-ReDoc-85ea2d?logo=openapiinitiative)](https://redocly.github.io/redoc/?url=HERE-URL-TO-RAW-FILE)
-->


## Related issue/s
<!-- LINK to other issues and add prefix `closes`, `fixes`, `resolves`-->
- closes https://github.com/ITISFoundation/osparc-simcore/issues/7635
- https://github.com/ITISFoundation/osparc-simcore/issues/6831

## How to test

<!-- Give REVIEWERS some hits or code snippets on how could this be tested -->

## Dev-ops
* https://github.com/ITISFoundation/osparc-ops-environments/pull/1040

GitHK commented on packages/service-library/src/servicelib/aiohttp/monitoring.py:
```diff
@@ -29,96 +28,13 @@
-kPROCESS_COLLECTOR = f"{__name__}.collector_process"  # noqa: N816
-kPLATFORM_COLLECTOR = f"{__name__}.collector_platform"  # noqa: N816
-kGC_COLLECTOR = f"{__name__}.collector_gc"  # noqa: N816
+kPROMETHEUS_METRICS: Final[str] = f"{__name__}.prometheus_metrics"  # noqa: N816
```
not C++, 🐍 !
```suggestion
_PROMETHEUS_METRICS: Final[str] = f"{__name__}.prometheus_metrics"  # noqa: N816
```

bisgaard-itis replied:
Sure, will fix

GitHK commented on packages/service-library/src/servicelib/fastapi/monitoring.py:
```diff
@@ -0,0 +1,128 @@
+    X_SIMCORE_USER_AGENT,
+)
+
+kPROMETHEUS_METRICS = "prometheus_metrics"
```
same here
```suggestion
_PROMETHEUS_METRICS = "prometheus_metrics"
```

bisgaard-itis replied:
good point. Will fix

GitHK commented on packages/service-library/src/servicelib/prometheus_metrics.py:
```diff
@@ -0,0 +1,220 @@
+    return {"TraceID": trace_id}
+
+
+def setup_prometheus_metrics() -> PrometheusMetrics:
```
setup name is a bit special. we now have `setup` and `initialize` and `lifespan` which would get mixed up with this one. Since you are not setting a module in fastapi please consider the below rename or something else that does not conclude the word "setup".
```suggestion
def get_prometheus_metrics() -> PrometheusMetrics:
```

GitHK reviewed (commented):
Thanks please consider my comments

matusdrobuliak66 reviewed (approved):
👍 thanks

sanderegg commented on packages/service-library/src/servicelib/aiohttp/monitoring.py:
```diff
@@ -1,25 +1,24 @@
 )
 from prometheus_client.registry import CollectorRegistry
 from servicelib.aiohttp.typing_extension import Handler
+from servicelib.prometheus_metrics import (
```
relative imports

bisgaard-itis replied:
thanks, will fix

sanderegg commented on packages/service-library/src/servicelib/fastapi/monitoring.py:
```diff
@@ -0,0 +1,136 @@
+from fastapi import FastAPI, Request, Response, status
+from fastapi_lifespan_manager import State
+from prometheus_client import CollectorRegistry
+from servicelib.prometheus_metrics import (
```
relative imports

bisgaard-itis replied:
thanks, will fix

sanderegg commented on packages/service-library/src/servicelib/fastapi/monitoring.py:
```diff
@@ -0,0 +1,136 @@
+
+
+def initialize_prometheus_instrumentation(app: FastAPI) -> None:
+    # NOTE: this cannot be ran once the application is started
```
can you add an assert or something then?

bisgaard-itis replied:
I will delete this comment. This is the same workflow for all the middleware we have, so I don't think the comment is needed anymore.

sanderegg commented on packages/service-library/src/servicelib/fastapi/monitoring.py:
```diff
@@ -0,0 +1,136 @@
+def initialize_prometheus_instrumentation(app: FastAPI) -> None:
+    # NOTE: this cannot be ran once the application is started
+
+    # NOTE: use that registry to prevent having a global one
```
that note does not mean anything now. where is the registry?

bisgaard-itis replied:
right, that comment came from a previous implementation. I will delete. Thanks

sanderegg commented on packages/service-library/src/servicelib/fastapi/monitoring.py:
```diff
@@ -0,0 +1,136 @@
+    app.add_event_handler("shutdown", _on_shutdown)
+
+    prometheus_metrics = app.state.prometheus_metrics
+    assert isinstance(prometheus_metrics, PrometheusMetrics)  # nosec
```
so correct me if I'm wrong, but it looks like you are re-implementing fastapi-instrumentation here? is this correct?

bisgaard-itis replied:
Hmm, I would rather say that I refactored the implementation we had for aiohttp so that I can use the same code for fastapi. In particular I removed the tool we used for fastapi (because it was insufficient).

sanderegg reviewed (commented):
Am I correct in thinking that you removed the fastapi-instrumentation library usage?
If yes should you not also remove its dependency?

YuryHrytsuk reviewed (approved):
> I decided to consolidate the aiohttp and fastapi middleware we use to collect prometheus metrics. 

What does it mean? This is not clear

bisgaard-itis commented:
> > I decided to consolidate the aiohttp and fastapi middleware we use to collect prometheus metrics.
> 
> What does it mean? This is not clear

Before we used our own custom implementation middleware for collecting prometheus metrics for aiohttp services and a pip installed tool for fastapi services. Now I have removed that tool and refactored the aiohttp middleware to have an aiohttp middleware and a separate fastapi middleware (custom implementation) which call the same module for collecting prometheus metrics

bisgaard-itis commented:
> Also I do not see any test

tests have been added.

sanderegg reviewed (approved):
thanks

pcrespov commented on packages/service-library/src/servicelib/fastapi/monitoring.py:
```diff
@@ -0,0 +1,136 @@
+        registry.unregister(collector)
+
+
+def setup_prometheus_instrumentation(app: FastAPI) -> CollectorRegistry:
```
please, create an alternative to this `setup` that supports [lifespan events](https://fastapi.tiangolo.com/advanced/events/) .  For example see  any of the `packages/service-library/src/servicelib/fastapi/*_lifespan.py` modules

bisgaard-itis replied:
This exists already below (see line 124). I don't know why there are two of these. I know it looks like I added this code from the diff, but I just moved it around, so I will not attempt to consolidate these two functions in this PR.

pcrespov commented on packages/service-library/tests/aiohttp/test_monitoring.py:
```diff
@@ -88,12 +89,12 @@ async def test_setup_monitoring(client: TestClient):
     response = await client.get("/metrics")
     assert response.status == status.HTTP_200_OK
     metrics_as_text = await response.text()
+    # metrics_as_text = metrics_as_text.lstrip("'").lstrip("'")
```
rm comment

bisgaard-itis replied:
done. Thanks for reminding

pcrespov reviewed (approved):
Very nice! I would like to ask you some specific details about it sometime offline
